File: 038-Longest Substring Without Repeating Characters/LongestSubstring.cpp
#include<iostream>
#include<unordered_map>
#include<string>
using namespace std;

int lengthOfLongestSubstring(string s){
    unordered_map<char, int> CharIndex;
    int maxLen = 0, start = 0;

    for(int end = 0; end < s.length(); end++){
        char Current = s[end];
        if(CharIndex.count(Current) && CharIndex[Current] >= start){
            start = CharIndex[Current] + 1;
        }
        CharIndex[Current] = end;
        maxLen = max(maxLen, end - start + 1);
    }
    return maxLen;
}

int main(){
    string s = "abcabcbb";
    cout<<"Length of Longest Substring: "<<lengthOfLongestSubstring(s)<<endl;
    return 0;
}

File: 038-Longest Substring Without Repeating Characters/LongestSubstring.py
def length_of_longest_substring(s: str) -> int:
    char_index = {}
    max_len = 0
    start = 0

    for end, char in enumerate(s):
        if char in char_index and char_index[char] >= start:
            start = char_index[char] + 1  # Move start after last occurrence
        char_index[char] = end
        max_len = max(max_len, end - start + 1)
    
    return max_len

# Example usage:
s = "abcabcbb"
print("Length of longest substring:", length_of_longest_substring(s))  # Output: 3